tasks/main.yml

```diff
@@ -4,11 +4,24 @@
     name: "{{ pip_package }}"
     state: present
 
+- name: Ensure virtualenv is installed if needed.
+  package:
+    name: "{{ pip_virtualenv_packages }}"
+    state: present
+  when: (pip_virtualenv is defined
+        or pip_install_packages | map(attribute='virtualenv') | select('defined') | length > 0)
+
 - name: Ensure pip_install_packages are installed.
   pip:
     name: "{{ item.name | default(item) }}"
     version: "{{ item.version | default(omit) }}"
-    virtualenv: "{{ item.virtualenv | default(omit) }}"
-    state: "{{ item.state | default(omit) }}"
-    executable: "{{ pip_executable }}"
+    state: "{{ item.state | default(pip_state | default(omit)) }}"
+    executable: "{{ pip_executable if item.virtualenv is not defined and pip_virtualenv is not defined else omit }}"
```

Comment:
This particular change seems like the only one I could see potentially causing an issue—the `else omit` part, couldn't that be triggered if `pip_virtualenv` is defined (but not necessarily desired)? (might just be my brain not working correctly with three conditions in one line ;).

Reply:
`virtualenv` parameter cannot be specified together with the `executable` parameter ([added in 2.1](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/pip_module.html#parameter-virtualenv)).
I guess this particular change is necessary to meet this constraint.

Reply:
'virtualenv' parameter cannot be specified together with the 'executable' parameter ([added in 2.1](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/pip_module.html#parameter-virtualenv)).
I guess this particular change is necessary to meet this constraint.

In my environment (Ansible 2.11.1), `else omit` is triggered when either `item.virtualenv` or `pip_virtualenv` is defined.

| `item.virtualenv` | `pip_virtualenv` | `executable` |
|:-:|:-:|:-:|
| defined | defined | `omit` |
| defined | undefined | `omit` |
| undefined | defined | `omit` |
| undefined | undefined | `pip_executable` |

Reply:
Thanks for putting in the work to understand what I did a few months ago 😬
Seems like it's does what it's supposed to do